lib/SILOptimizer/Mandatory/MoveOnlyAddressCheckerUtils.cpp

```diff
@@ -1737,6 +1770,12 @@ bool GatherUsesVisitor::visitUse(Operand *op) {
     LLVM_DEBUG(llvm::dbgs() << "Running copy propagation!\n");
     moveChecker.changed |= moveChecker.canonicalizer.canonicalize();
 
+    // Export the drop_deinit's discovered by the ObjectChecker into the
+    // AddressChecker to preserve it for later use. We need to do this since
+    // the ObjectChecker's state gets cleared after running on this LoadInst.
+    for (auto *dropDeinit : moveChecker.canonicalizer.getDropDeinitUses())
```

Comment:
Can you put in an assert if we see a drop_deinit on an address. That way when we loosen the restriction and allow for discard to be used in more contexts, we know that we need to update this code.

Reply:
So at this point, if we did export a drop_deinit on an address, it's fine, because we later will only use the presence of the instruction to determine diagnostics. I think what you're asking for is a check that we didn't miss a drop_deinit of an address, because the ObjectChecker is not going to see those, and thus we would fail to diagnose consume-on-all-paths. 

I can add a not-NDEBUG check for this in a follow-up PR that stays on `main`. I didn't do it here because it's rather involved: It'd amount to scanning scanning every consuming method to look for these instructions (or adding a tag in the SILFunction / FuncDecl). In addition, because the move checker runs this same checking on more than just `self`, we'd then need to disambiguate between what we're doing checking on. Since we don't allow discard in noncopyables with non-trivial fields, it can miss 5.9

Reply:
That is not what I was thinking about specifically. What I was imagining was adding a quick assert into the UseGatherer and assert if we see it. That being said, I thought about this further and realized that we can have drop_deinit for addresses (specifically in deinits of address only noncopyable types). Of course this isn't as a result of running discard so we don't want to check it (and since we do not allow reinits in deinits for now, we can't have code after the final drop_deinit). So putting in an assert would actually cause us to assert.

So ignore this.